README.md
[![Travis Build Status](https://img.shields.io/travis/mstrimas/prioritizr/master.svg?label=Mac%20OSX%20%26%20Linux)](https://travis-ci.org/mstrimas/prioritizr)
[![AppVeyor Build Status](https://img.shields.io/appveyor/ci/mstrimas/prioritizr/master.svg?label=Windows)](https://ci.appveyor.com/project/mstrimas/prioritizr)
[![Coverage Status](https://codecov.io/github/mstrimas/prioritizr/coverage.svg?branch=master)](https://codecov.io/github/mstrimas/prioritizr?branch=master)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/prioritizr)](https://CRAN.R-project.org/package=prioritizr)

<!-- README.md is generated from README.Rmd. Please edit that file -->



# Conservation Prioritization with Integer Programming in R

`prioritizr` is an R package for solving systematic conservation prioritization problems using the techniques of integer linear programming (ILP). In particular, both the minimum set cover (Marxan-like) and maximum coverage reserve design problems can be solved. The package offers a unified interface to a variety of commercial and open-source ILP solvers. In contrast to heuristic methods, such as simulated annealing, the ILP methods used by `prioritizr` can find exact solutions to optimization problems.

This package consists largely of two layers of functions: those that define a reserve design problem and those that solve a reserve design problem using any one of a variety of solvers. The details of the solvers are intentionally abstracted away so that the user requires minimal knowledge of the specific solvers or ILP in general.

The currently supported solvers are as follows. Each must be installed separately from this package to be accessible.

- [Gurobi](http://gurobi.com) is a state-of-the-art commercial optimization software with an R package interface. It is by far the fastest of the solvers available in this package, however, it is also the only one that isn't free. That said, free academic licenses are available.
- [SYMPHONY](https://projects.coin-or.org/SYMPHONY) is an open-source integer programming solver that is part of the Computational Infrastructure for Operations Research (COIN-OR) project, an initiative to promote development of open-source tools for operations research (a field that includes linear programming). Two R packages exist to provide interfaces to
SYMPHONY: `Rsymphony` (on CRAN) and `lpsymphony` (on Bioconductor). On Windows and Mac, `lpsymphony` may be easier to install.
- The GNU Linear Programming Kit ([GLPK](https://www.gnu.org/software/glpk/)) is an open-source package for solving linear and integer linear programs. The R package
glpkAPI provides an interface to the low-level GLPK API.

# Installation


```r
if (packageVersion("devtools") < 1.6) {
  install.packages("devtools")
}
devtools::install_github("mstrimas/prioritizr")
```

# Example usage

For full details on this package, consult the vignette with `browseVignettes("prioritizr")` or the [package documentation](http://mstrimas.github.io/prioritizr). What follows are some simple example of how this package can be used to solve conservation prioritization problems.

## Required Packages


```r
library(prioritizr)
# for plotting
library(rasterVis)
library(viridis)
```

## Data generation

The function `gaussian_field()` can be used to generate spatially auto-correlated random fields, to be used as semi-realistic spatial variables. All conservation prioritization problems involve balancing a trade-off between representing features of conservation interest and minimizing the cost of protection. Therefore we start by generating distributions for four species and a cost layer, all on a 100x100 raster grid of planning units.


```r
set.seed(1)
# raster 100x100 template
e <- raster::extent(0, 100, 0, 100)
r <- raster::raster(e, nrows = 100, ncols = 100, vals = 1)
# generate 9 feature distributions with different scales and range sizes
species <- mapply(function(x, y, r) gaussian_field(r = r, range = x, prop = y),
            rep(c(5, 25), each = 2),
            rep(c(0.1, 0.5), times = 2),
            MoreArgs = list(r = r))
species <- raster::stack(species)
species <- setNames(species, letters[1:raster::nlayers(species)])
levelplot(species, main = 'Species Distributions', layout = c(2, 2),
          scales = list(draw = FALSE),
          col.regions = c("grey20", "#fd9900"), colorkey = FALSE)
```

![plot of chunk generate-data](README/generate-data-1.png)

```r
# genrate cost layer
cost <- gaussian_field(r, 20, mean = 1000, variance = 500)
cost <- setNames(cost, "cost")
levelplot(cost, main = "Cost", margin = FALSE,
          col.regions = viridis::viridis(100))
```

![plot of chunk generate-data](README/generate-data-2.png)

## Minimum set cover problem

In the context of systematic reserve design, the minimum set cover problem seeks to find the set of planning units that minimizes the overall cost of a reserve network, while meeting a set of representation targets for the conservation features. The cost is often either the area of the planning units or the opportunity cost of foregone commercial activities (e.g. logging or agriculture). The representation targets ensure that each species is adequately represented in the reserve network.

This problem is equivalent to a simplified Marxan reserve design problem,
with the Boundary Length Modifier (BLM) set to zero. To specify a prioritization model of this type we use the `minsetcover_model()` function to create a `minsetcover_model` S3 object. This function takes data in a variety of formats (raster, vector, or tabular) and generates a standard object encapsulating the prioritization problem. Here I set targets such that 20% of each species' existing range will be protected.


```r
msc_model <- minsetcover_model(x = cost, features = species, targets = 0.2)
class(msc_model)
#> [1] "minsetcover_model" "prioritizr_model"
```

## Maximum coverage problem

The maximum coverage problem seeks to find the set of planning units that maximizes the overall level of representation across a suite of conservation features, while keeping cost within a fixed budget. The cost is often either the area of the planning units or the opportunity cost of foregone commercial activities (e.g. from logging or agriculture). Representation within each planning unit is typically given by the occupancy of each species, however, some measure of abundance or probability of occurrence may also be used.

This problem is roughly the opposite of what the conservation planning software Marxan does. To specify a prioritization model of this type we use the `maxcover_model()` function to create a `maxcover_model` S3 object. This function takes data in a variety of formats (raster, vector, or tabular) and generates a standard object encapsulating the prioritization problem. Here we set the budget to 25% of the total cost of the study area .


```r
b_25 <- 0.25 * raster::cellStats(cost, "sum")
mc_model <- maxcover_model(x = cost, features = species, budget = b_25)
class(mc_model)
#> [1] "maxcover_model"   "prioritizr_model"
```

## Maximum target coverage problem

The maximum target coverage problem is a modified version of the the maximum coverage problem. Each conservation feature is assigned a representation target and the objective is to find the set of planning units that meets the most targets while remaining within a fixed budget. 

This problem is meant to be a hybrid between the maximum coverage problem and a Marxan-like minimum set cover problem in that it allows for both a budget and targets to be set. To specify a prioritization model of this type we use the `maxtargets_model()` function to create a `maxtargets_model` S3 object. This function takes data in a variety of formats (raster, vector, or tabular) and generates a standard object encapsulating the prioritization problem. Here we set the budget to 10% of the total cost of the study area and choose targets such that 20% of each species' existing range will be protected.


```r
b_10 <- 0.1 * raster::cellStats(cost, "sum")
mtc_model <- maxtargets_model(x = cost, features = species, targets = 0.2, 
                           budget = b_25)
class(mtc_model)
#> [1] "maxtargets_model" "prioritizr_model"
```

## Solving prioritization problems

The function `prioritize()` offers a unified interface to solving either type of prioritization problem using any of the available solvers. To solve the minimum set cover problem to within 0.1% of optimality use:


```r
msc_results <- prioritize(msc_model, gap = 0.001)
```

By default, this function uses the best available solver, which is Gurobi if it is installed. Alternatively, the solver can be specified explicitly. For example, to solve the maximum coverage problem with SYMPHONY use:


```r
mc_results <- prioritize(mc_model, solver = "symphony", gap = 0.001)
```

And to solve the maximum target coverage problem with GLPK use:


```r
mtc_results <- prioritize(mtc_model, solver = "glpk", gap = 0.001)
```

The resulting solution can be displayed with `plot_selection()`:


```r
plot_selection(cost, msc_results$x, title = "Minimum Set Cover Solution")
```

![plot of chunk solutions](README/solutions-1.png)

```r
plot_selection(cost, mc_results$x, title = "Maximum Coverage Solution")
```

![plot of chunk solutions](README/solutions-2.png)

```r
plot_selection(cost, mtc_results$x, title = "Maximum Target Coverage Solution")
```

![plot of chunk solutions](README/solutions-3.png)

Finally, the objective function values for these solutions can be access with:


```r
# minimum set cover objective
msc_results$objval
#> [1] 985393.9
# maximum coverage objective
mc_results$objval
#> [1] 5667
# maximum target coverage objective
mtc_results$objval
#> [1] 3.998998
```


NEWS.md
# prioritizr 0.1.2 (unreleased)

* prepare for CRAN submission
* add continuous integration
* fixed various bugs
* use target-based formulation of maximum coverage problem

# prioritizr 0.1.1 (unreleased)

* Initial package version



R/prioritize.r
#' Solve a conservation prioritization problem using ILP
#'
#' Solve a systematic conservation planning problem using the the methods of
#' Integer Linear Programming (ILP). In particular, this function solves either
#' the the minimum set cover problem or the maximum coverage problem for
#' systematic reserve design. In addition, this function provides a unified
#' interface to a variety of optimization packages and gives the user the
#' ability to chose which to use based on the availability on their machine.
#' Choosing a solver is done via the \code{solver} argument of the
#' \code{prioritize()} function, or by invoking one of the \code{prioritize_*}
#' functions.
#'
#' @section Prioritization problems:
#'
#' The \code{prioritizr} package currently handles three types of conservation
#' prioritization problems. Each type of problem has a corresponding
#' \code{prioritizr_model} object that encapsulates the problem and is passed
#' as the first argument to \code{prioritize()}:
#'
#' \itemize{
#'   \item \bold{Minimum set cover}: find the set of planning units that
#'   minimizes the overall cost of a reserve network, while meeting a set of
#'   representation targets for the conservation features. The cost is often
#'   either the area of the planning units or the opportunity cost of foregone
#'   commericial activities (e.g. logging or agriculture). The representation
#'   targets ensure that each species is adequately represented in the reserve
#'   network.
#'
#'   This problem is equivalent to a simplified Marxan reserve design problem,
#'   with the Bounday Length Modifier (BLM) set to zero. Use
#'   \code{\link{minsetcover_model}} to construct a minimum set cover model.
#'
#'   \item \bold{Maximum coverage}: find the set of planning units that
#'   maximizes the overall level of representation across a suite of
#'   conservation features, while keeping cost within a fixed budget. The cost
#'   is often either the area of the planning units or the opportunity cost of
#'   foregone commericial activities (e.g. logging or agriculture).
#'   Representation level is typically given by the occupancy within each
#'   planning unit, however, some measure of abundance or probability of
#'   occurence may also be used.
#'
#'   This problem is roughly the opposite of what the conservation planning
#'   software Marxan does. Use \code{\link{maxcover_model}} to construct a
#'   maximum coverage model.
#'
#'   \item \bold{Maximum coverage of targets}: this is a modified version of
#'   the maximum coverage problem. Each conservation feature is assigned a
#'   representation target and the objective is to find the set of planning
#'   units that meets most targets while remaining within a fixed budget.
#'
#'   This problem is meant to be a hybrid between the maximum coverage problem
#'   and a Marxan-like minimum set cover problem in that it allows for both a
#'   budget and targets to be set. Use \code{\link{maxtargets_model}} to
#'   construct a maximum target coverage model.
#' }
#'
#' @section Solvers:
#'
#' The following optimization packages (and corresponding values for the
#' \code{solver} parameter) are supported. The packages must be installed
#' independently to function:
#'
#' \itemize{
#'   \item \bold{\code{gurobi}}: \href{http://gurobi.com}{Gurobi} is a
#'   state-of-the-art commercial optimization software with an R package
#'   interface. It is by far the fastest of the solvers available in this
#'   package, however, it is also the only one that isn't free. That said, free
#'   academic licenses are available.
#'   \item \bold{\code{symphony}}:
#'   \href{https://projects.coin-or.org/SYMPHONY}{SYMPHONY} is an open-source
#'   integer programming solver that is part of the Computational Infrastructure
#'   for Operations Research (COIN-OR) project, an initiative to promote
#'   development of open-source tools for operations research (a field that
#'   includes linear programming). Two R packages exist to provide interfaces to
#'   SYMPHONY: \code{Rsymphony} (on CRAN) and \code{lpsymphony} (on
#'   Bioconductor). On Windows and Mac, \code{lpsymphony} may be easier to
#'   install. \code{prioritize()} will choose whichever package is available if
#'   \code{solver = "symphony"} is used.
#'   \item \bold{\code{glpk}}: The GNU Linear Programming Kit
#'   (\href{https://www.gnu.org/software/glpk/}{GLPK}) is an open-source
#'   package for solving linear and integer linear programs. The R package
#'   glpkAPI provides an interface to the low-level GLPK API.
#' }
#'
#' @param pm \code{prioritizr_model} object specifying the prioritization model
#'   to solve. This will typically be output from one of the corresponding
#'   functions: \code{\link{minsetcover_model}} or \code{\link{maxcover_model}}.
#' @param solver character; specify the optimization package to use to solve the
#'   prioritization problem. Availability will depend on what packages are
#'   installed, however, the full list of possibilities is: gurobi, symphony,
#'   and glpk. Alternatively, specify "best" to automatically choose the best
#'   optimizer currently installed.
#' @param gap numeric; the relative gap to optimality. The optimizer will
#'   terminate when the difference between the upper and lower objective
#'   function bounds is less than the gap times the upper bound. For example, a
#'   value of 0.01 will result in the optimizer stopping when the difference
#'   between the bounds is 1 percent of the upper bound.
#' @param time_limit numeric; time limit in seconds to run the optimizer. The
#'   solver will return the current best solution when this time limit is
#'   exceeded.
#' @param first_feasible logical; whether to return the first feasible solution.
#'   If \code{first_feasible} is set to \code{TRUE}, the solver will return the
#'   first solution it encounters that meets all the constraints, regardless of
#'   solution quality. Note that the first feasible solution is not an arbitrary
#'   solution, rather it is derived from the relaxed solution, and is therefore
#'   often reasonably close to optimality.
#' @param bound numeric; lower bound on the objective function. To return the
#'   gap to optimality, the solver requires a lower bound on the objective
#'   function. Gurobi will estimate this lower bound, however, most other
#'   solvers won't, in which case, this argument can be supplied if the bound is
#'   known from some other source. This value is only used to calculate the
#'   optimality gap and will have no effect on the operation of the solver.
#' @param threads non-negative integer; the number of threads to use for the
#'   optimization algorithm. The default value of 0 will result in all cores in
#'   the machine being used. Currently only implemented for Gurobi.
#'
#' @return A \code{prioritizr_results} object containing the solution to the
#'   prioritization problem. This is an S3 object consisting of a list with the
#'   following components:
#'
#' \itemize{
#'   \item \code{x}: vector of decision variables for the best solution. 1
#'     indicates that a planning unit is selected, 0 that it isn't.
#'   \item \code{objval}: objective function value for optimal solution.
#'   \item \code{objbound}: lower bound on objective function. This is useful
#'     only if the solver was stopped before finding the optimum. If the solver
#'     was run to completion, \code{objbound} and \code{objval} will be equal.
#'   \item \code{gap}: the relative gap to optimality:
#'     \code{objval} / \code{objbound} - 1. \code{gap} will be zero if the
#'     solver was run to comletion and the true optimum was found.
#'   \item \code{time}: the execution time (in seconds) of the solver.
#' }
#'
#' @export
#' @examples
#' # raster 100x100 template
#' e <- raster::extent(0, 100, 0, 100)
#' r <- raster::raster(e, nrows = 100, ncols = 100, vals = 1)
#'
#' # generate 9 feature distributions with different scales and range sizes
#' f <- mapply(function(x, y, r) gaussian_field(r = r, range = x, prop = y),
#'             rep(c(5, 15, 25), each = 3),
#'             rep(c(0.1, 0.25, 0.5), times = 3),
#'             MoreArgs = list(r = r))
#' f <- raster::stack(f)
#' f <- setNames(f, letters[1:raster::nlayers(f)])
#' # genrate cost layer
#' cost <- gaussian_field(r, 20, mean = 1000, variance = 500)
#' cost <- setNames(cost, "cost")
#'
#' # prepare minimum set cover prioritization model
#' # use 20% targets
#' msc_model <- minsetcover_model(x = cost, features = f, targets = 0.2)
#' # solve to within 1 percent of optimality
#' # pick solver automatically (uses Gurobi if installed)
#' msc_results <- prioritize(msc_model, gap = 0.001)
#' plot_selection(cost, msc_results$x)
#' # specify SYMPHONY solver
#' msc_results_symphony <- prioritize(msc_model, solver = "symphony",
#'                                    gap = 0.001)
#' plot_selection(cost, msc_results_symphony$x)
#'
#' # prepare maximum coverage prioritization model
#' # set budget to 25% of total cost
#' b_25 <- 0.25 * raster::cellStats(cost, "sum")
#' mc_model <- maxcover_model(x = cost, features = f, budget = b_25)
#' # solve to within 1 percent of optimality
#' # pick solver automatically (uses Gurobi if installed)
#' mc_results <- prioritize(mc_model, gap = 0.001)
#' plot_selection(cost, mc_results$x)
#'
#' # prepare maximum target coverage prioritization model
#' # set budget to 25% of total cost
#' b_25 <- 0.25 * raster::cellStats(cost, "sum")
#' mtc_model <- maxtargets_model(x = cost, features = f, budget = b_25,
#'                             targets = 0.5)
#' # solve to within 1 percent of optimality
#' # pick solver automatically (uses Gurobi if installed)
#' mtc_results <- prioritize(mtc_model, gap = 0.001)
#' plot_selection(cost, mtc_results$x)
prioritize <- function(pm,
                       solver = c("best", "gurobi", "symphony", "glpk"),
                       gap = 1e-4,
                       time_limit = Inf,
                       first_feasible = FALSE,
                       bound = NA_real_,
                       threads = 0) {
  # assertions on arguments
  solver <- match.arg(solver)
  assert_that(inherits(pm, "prioritizr_model"),
              assertthat::is.number(gap),
              gap >= 0,
              assertthat::is.number(time_limit),
              time_limit > 0,
              assertthat::is.flag(first_feasible),
              assertthat::is.number(bound),
              is_integer(threads), length(threads) == 1)
  # choose best solver
  if (solver == "best") {
    if (requireNamespace("gurobi", quietly = TRUE)) {
      solver <- "gurobi"
    } else if (requireNamespace("Rsymphony", quietly = TRUE)) {
      solver <- "symphony"
    } else if (requireNamespace("lpsymphony", quietly = TRUE)) {
      solver <- "symphony"
    } else if (requireNamespace("glpkAPI", quietly = TRUE)) {
       solver <- "glpk"
    } else {
      stop("No ILP solvers installed.")
    }
  }
  # pick solver
  if (solver == "gurobi") {
    pr <- prioritize_gurobi(pm = pm, gap = gap, time_limit = time_limit,
                            first_feasible = first_feasible, bound = bound,
                            threads = threads)
  } else if (solver == "symphony") {
    pr <- prioritize_symphony(pm = pm, gap = gap, time_limit = time_limit,
                              first_feasible = first_feasible, bound = bound)
  } else if (solver == "glpk") {
    if (first_feasible) {
      stop("first_feasible = FALSE not permitted for GLPK solver.")
    }
    pr <- prioritize_glpk(pm = pm, gap = gap, time_limit = time_limit,
                          bound = bound)
  } else {
    stop("Invalid solver.")
  }
  return(pr)
}
